### DISCUSSION TITLE
Whither custom self-labels?

### DISCUSSION THREAD

**Discussion by snarfed** [bluesky-social/atproto#2885] [Protocol (atproto)]
Hi all! What's the current status and thinking on custom (ie non-bsky.app) [self-labels](https://atproto.com/specs/label#self-labels-in-records)? They are usable right now, granted; Bridgy Fed sets one on each profile bridged into Bluesky from the fediverse. However, unlike labels served by a mod service, it seems like there isn't a straightforward path to expose self-labels to users? https://atproto.com/specs/label#self-labels-in-records says "downstream clients" can use them. Is that the expectation right now: clients have to surface them explicitly if they want to? Any plans for bsky.app to do that? Any other thoughts?

More background in https://github.com/snarfed/bridgy-fed/issues/1388 , cc @qazmlp . Thanks in advance!

**Comment by snarfed** [comment]
One approach to this could be a generic self-labeler mod service that just re-emits all self-labels. Seems simple enough, should do the trick, and anyone could build it...?

**Reply by qazmlp** [reply]
I'm not sure that would work, in terms of displaying custom ones in a human-friendly way.
The self-labels themselves don't seem to contain enough information to do that properly.

**Reply by snarfed** [reply]
True! The self-label mod service should ideally come up with human-friendly names for all the self-labels it knows about and publish those in its `app.bsky.labeler.service` https://docs.bsky.app/docs/advanced-guides/moderation#labeler-declarations . It could maybe take a first pass on its own, and then offer to cooperate with the original self-label authors to accept revisions.

**Reply by snarfed** [reply]
If I built this, I'd probably also ignore bsky.app's self labels too, since they're already displayed, and to minimize load. I'm more interested in surfacing truly custom self labels that don't currently get displayed at all.

**Reply by snarfed** [reply]
I'm also curious how realistic it'd be to make this online-only, ie not store labels, just apply them on the fly.

https://atproto.com/specs/label#label-distribution-endpoints says labelers need to serve `com.atproto.label.subscribeLabels` and `queryLabels`. `subscribeLabels` seems fine, just consume the firehose or jetstream and emit self labels as it sees them, except for `cursor`, I'd probably avoid that. `queryLabels` is doable too, loading full `at://` URI records synchronously is fine, but I'd be reluctant to handle patterns too, eg load all posts in a given repo, or all records.

I'm curious about the API calls that the AppView makes to labelers in practice. Does it ever use patterns? Which kinds? How often? etc.

**Reply by snarfed** [reply]
Plot twist! [Jaz says in Discord](https://discord.com/channels/1097580399187738645/1097580399187738648/1296972640938102814), "the appview never calls queryLabels, it only relies on the websocket."

Definitely simplifies this! I'm sorely tempted to ignore Bridgy Fed for a bit and crank this out. 😳

**Reply by qazmlp** [reply]
> Plot twist! [Jaz says in Discord](https://discord.com/channels/1097580399187738645/1097580399187738648/1296972640938102814), "the appview never calls queryLabels, it only relies on the websocket."
> 
> Definitely simplifies this! I'm sorely tempted to ignore Bridgy Fed for a bit and crank this out. 😳

The caveat being that no new AppView could backfill those if they can't be queried, at least without trusting Bluesky.

I think it's important to implement `queryLabels` in terms of not locking-in the ATProto ecosystem to Bluesky-the-AppView-instance.

**Reply by snarfed** [reply]
True! Although the team's AppView itself couldn't backfilll either, for records created before the self-labeler launches. So a quantitative difference, but maybe not a qualitative one.

...although having said that, I don't know that third parties are expected to run AppViews for lexicons they don't own...? The AppView implements the app and lexicon semantics, and those are generally owned by a single entity for a given app, eg the Bluesky team for bsky.app. I _think_ https://bsky.social/about/blog/5-5-2023-federation-architecture says as much:

> The federation architecture allows anyone to host a Relay...There will also be an ecosystem of App Views for each lexicon, or “social mode,” deployed on the network. For example, Bluesky currently supports a micro-blogging mode: the app.bsky lexicon. Developers who create new lexicons would likely deploy a corresponding App View that understands their lexicon to service their users.

**Reply by snarfed** [reply]
Unrelated, @qazmlp you've clearly been interested and (very!) knowledgeable about decentralized social and ATProto and building software in general, but haven't had the free time to take on a big project...? This one seems pretty small and contained. Seems like it might fit you well?

**Reply by qazmlp** [reply]
@snarfed I have several equally 'small' ongoing projects that I'm already not getting to 😅 (Sorry.)

I know it's not necessarily good form to only discuss and not publish any code, but this really is the extent I can spare of myself in the ATProto context.

**Reply by bnewbold** [reply]
> I don't know that third parties are expected to run AppViews for lexicons they don't own...?

It is a very important part of the protocol/project that it be possible (and *credible*) for a third party to step in and start running an appview for any app/Lexicon. More broadly, any infrastructure that any community relies on should have "adversarial interop", meaning a new provider could replace existing providers (if they can convince users to jump over, and have some realistic/sustainable funding model). We think it would also be good for there to be multiple providers for major/important/popular app modalities, but it isn't as high of a priority right now as the "credible" bit, and things like improving self-control of identity and making new app development easier (gotta prioritize!).

In general I do think it is important for folks to be able to backfill labels. Note that at least the Ozone mod service allows full backfill on the subcribeLabels endpoint, so both subscribeLabels and queryLabels could be used for backfill. When Jaz said that the AppView doesn't use queryLabels, I think they were just describing how the system works today. We think queryLabels is a fairly important part of the API surface going forward (it enables a lot of use cases, introspection, etc), even if we aren't currently using it. And hopefully isn't too laborious to implement.

(of course, as always, folks can experiment as they see fit; my role here is to nudge in the direction of certain norms)

**Reply by snarfed** [reply]
Thanks for the correction @bnewbold! Sorry for unintentionally spreading misinformation. 😁

One big part of what led me astray here was the question of lexicon and app "ownership." I was maybe thinking more about whether third parties could or should _build_ appviews for lexicons/apps they don't own, beyond merely running independent instances of the reference implementation. Any nontrivial app will have nontrivial product semantics and behavior that are built into their appview. Would third party appviews be expected to try to match those semantics? Or is part of appview adversarial interop that they could offer alternative semantics? Either way could be reasonable, or maybe other ways entirely, but the answer obviously impacts what it means to "own" an app or set of lexicons, and what the ecosystem norms and expectations are.

(Apologies for the tangent from the topic here, happy to move somewhere else if you know where!)

**Reply by snarfed** [reply]
> In general I do think it is important for folks to be able to backfill labels.

Definitely understood, and agreed. I only interpreted Jaz as describing the current state, not being normative. I ended up nerd sniping myself into actually building this self-labeler, it's up and running and almost working, https://bsky.app/profile/self-labeler.snarfed.org , but I don't know if I plan to go the extra mile to make it robust or long-lived or backfill-able. We'll see!

**Comment by snarfed** [comment]
(If you're wondering whether the real reason I filed this was just so I could use the word _whither_, the answer is a resounding yes. English is ~~fun~~ ~~ridiculous~~ fun! 🤪)

**Comment by bnewbold** [comment]
In the big picture, we know that the labeling system probably needs a "v2" iteration. This probably won't be breaking protocol changes, but there are a large list of lessons learned and problems that have surfaced. Unfortunately we probably aren't going to be making smaller tweaks or improvements until we do that large push. This isn't to say that composable moderation isn't a priority.

Self-labels kind of fall in that bucket. As designed, self-labels are in the scope/context of the author, regardless of which labelers the viewer subscribes to. To have such labels appear consistently, we felt it was important that they be defined at the app level, not at labeler-service level. We also wanted to start with a small set of initial self-labels to start. We wanted to avoid a norm of "content warning policing", where folks are shamed or punished for not self-labeling.

All those decisions might be re-visited now that we have more experience! The "badging" use-case of labels has been very popular. But also having hashtags on profiles could cover many of those use-cases? Or other forms of badging, maybe community-specific, maybe tied to social roles, maybe something like https://docs.smokesignal.events/blog/blue-badge-released/? Lots of options to consider.

**Reply by qazmlp** [reply]
I'd only be half-joking if I said I think people want forum signatures.

Have you considered a sort of "micro bio" that's a little deemphasised next to the user name?
![image](https://github.com/user-attachments/assets/914cbb2c-bd7f-4bf5-8083-566f2bd1a426)

This is similar to how the Mastodon app (Moshidon) on my phone displays the "Pronouns" profile field.
It looks like UGC so I think there are fewer safety issues than with badges (but you'd likely still need _some_ filter for it).

I imagine people would want to use this for their fursona species or job title in some cases, too, like "internet goat" or "UX is my hobby". You may want to force text presentation for emoji though.

**Comment by snarfed** [comment]
@futurGH raised a great point just now: do custom self labels not actually exist? Or more specifically, does the protocol not allow them? https://docs.bsky.app/docs/advanced-guides/moderation#global-label-values says:
> The first is because only label values which are defined globally can be used as self-labels (ie set by a user who is not a Labeler). The porn, sexual, gore, nudity, and !no-unauthenticated labels are global for this reason.

Whoa! News to me. Seems kind of the opposite in spirit to the idea that clients can add arbitrary fields to records that aren't in their lexicon, and PDSes etc are required to preserve those fields, right? Even if we're now moving toward recommending sidecar records instead.

...but if this is true, then arguably I've been doing it wrong all this time by setting `bridged-from-bridgy-fed-activitypub` and `bridged-from-bridgy-fed-web` as self-labels on `app.bsky.actor.profile`s, and I should instead have been running my own labeler to emit those labels.

Phew, ok. If anyone can confirm, please do!

**Reply by qazmlp** [reply]
Somewhat related: https://github.com/bluesky-social/social-app/issues/3872

However, I do not know anything about the technical details and whether self-labels would show up in the app if they're attributed to a third party labeller.

**Comment by snarfed** [comment]
Finally found a custom self-label that's not from Bridgy Fed! `〔>_<〕`, on [at://did:plc:clacojbcivzii7ot2vdkt574/app.bsky.actor.profile/self](https://atproto-browser.vercel.app/at/did:plc:clacojbcivzii7ot2vdkt574/app.bsky.actor.profile/self) 😆

**Reply by snarfed** [reply]
Saw a few more custom self-labels this morning, all Japanese, eg https://atproto-browser.vercel.app/at/did:plc:lmftezsq52hi53taz762s7pc/app.bsky.actor.profile/self

**Comment by bleonard252** [comment]
Currently self labels in Bluesky aren't terribly useful since they have to have a global definition to define the rules, display information, and such. It may be helpful or even necessary for custom self-labels to have an extra field referring to a separate definition record, perhaps only accepting one that explicitly allows self-labeling. (The use case for this is that I'd like for people to be able to self-label USPol posts, using my labeler's labels so that any settings a reader already has for the label apply.)

**Reply by bleonard252** [reply]
Theoretically, it would be already possible and legal in the protocol to add a labeler record to a user, since self-labels always refer to the poster as the issuer? Again, it seems less useful to me since binding custom self-labels to a labeler means each label can be configured by an end-user as they see fit.

**Reply by yamarten** [reply]
I agree with the usefulness of having a self-label definition record. I expect it to be added with label v2.

However, it may not completely replace the self-label labeler. Simple self-labels are defined independently for each user, so viewers need to set the display style for each. For example, if a self-label labeler can provide a "semi-global label" with a common meaning, viewers can use it more conveniently.

It is probably possible to include definition sharing options in the bsky self-label system itself, but it is natural for me to use the labeler as an authority for label definition.
